parse daemon's response for errors

enoodle opened the pull request:
This will parse the response for errors and will attempt the next authentication configuration if an error is found.

To easily communicate the errors found a new channel is created and the go routines that handles the parsing and reporting will be started and destroyed for each authentication attempt.

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1449742

simon3z commented:
@pweil- can you review?

enoodle commented:
@ilackarms can you review?

simon3z commented:
Ping @pweil- @ilackarms this is pretty urgent. Can you review?

pweil- commented:
sorry, I am not getting any GH notifications for some reason.  Will review this today.

pweil- commented on pkg/inspector/image-inspector.go:
```diff
@@ -166,27 +166,40 @@ func aggregateBytesAndReport(bytesChan chan int) {
 			break
 		}
 		// decoding
+		if v.Error != "" {
```
what writes to the channel in the case of no decode error and no `v.Error`?  Is that going to block the call below of `if parsedError := <-parsedErrors; parsedError {` forever?  Lots going on here, I couldn't quite tell if that's a valid use case (will pull this PR today and see if I can reproduce anything)

pweil- commented on pkg/inspector/image-inspector.go:
```diff
@@ -210,27 +223,33 @@ func (i *defaultImageInspector) pullImage(client *docker.Client) error {
+		if authErr := client.PullImage(imagePullOption, auth); authErr != nil {
+			return authErr
+		}
+		writer.Close()
```
strange here that you're deferring reader and writer close above but explicitly closing them here.  Why?

pweil- commented on pkg/inspector/image-inspector.go:
```diff
@@ -210,27 +223,33 @@ func (i *defaultImageInspector) pullImage(client *docker.Client) error {
+		writer.Close()
+		reader.Close()
+		if parsedError := <-parsedErrors; parsedError {
+			log.Printf("Authentication with %s failed: %v", name, authErr)
```
auth error will always be nil according the the `client.PullImage` check above

pweil- commented on pkg/inspector/image-inspector_test.go:
```diff
@@ -165,6 +166,37 @@ func TestGetAuthConfigs(t *testing.T) {
+		defer reader.Close()
+		go decodeDockerPullMessages(parsedErrors, bytesChan, reader)
+		writer.Write([]byte(test_params.readerInput))
+		writer.Close()
```
closing after defer again?

pweil- commented on pkg/inspector/image-inspector.go:
```diff
@@ -166,27 +166,40 @@ func aggregateBytesAndReport(bytesChan chan int) {
+		Error          string
 	}
 	defer func() { close(bytesChan) }()           // Closing the channel to end the other routine
+	defer func() { close(parsedErrors) }()        // Close channel
```
closing the channels here rather than allowing the creator to do it is suspicious

pweil- commented on pkg/inspector/image-inspector.go:
```diff
@@ -166,27 +166,40 @@ func aggregateBytesAndReport(bytesChan chan int) {
 		if err := dec.Decode(&v); err != nil {
 			if err != io.ErrClosedPipe {
 				log.Printf("Error decoding json: %v", err)
+			} else {
```
maybe we should consider an error channel instead a bool channel, any decoding errors will not be displayed

pweil- commented on pkg/inspector/image-inspector.go:
```diff
@@ -210,27 +223,33 @@ func (i *defaultImageInspector) pullImage(client *docker.Client) error {
+		}
+		writer.Close()
+		reader.Close()
+		if parsedError := <-parsedErrors; parsedError {
```
the way that `decodeDockerPullMessages` is coded seems to indicate that decoding can happen many times and that multiple calls to the `parsedErrors` channel may happen.  I think that more strongly indicates that either we need to aggregate errors in `decodeDockerPullMessages` or this would need to change.  

It has been a while since I've been in this code so maybe this is a bad question: do we know why the `v.Error` doesn't surface in the `client.PullImage` call since it returns an error?  It seems weird that we'd have to have both the error check and parsing the progress stream for errors.

enoodle commented:
@pweil- can you take another look? Still closing the pipe manually, but I fixed other comments.
I am not sure how to signal to the decoding function that no more messages might come. I though using an input channel but then in the case that it takes long time to decode the messages it might cause us to skip some.

pweil- commented on pkg/inspector/image-inspector_test.go:
```diff
@@ -165,6 +166,43 @@ func TestGetAuthConfigs(t *testing.T) {
+		expectedErrors bool
+	}{
+		"no error": {readerInput: no_error_input, expectedErrors: false},
+		"error":    {readerInput: one_error, expectedErrors: true},
```
test for decode error too.  Also, please check to make sure you're getting the message you think you are.  I had an issue with some of my code where the test errored correctly but it was never hitting the code I thought it was...oops :smile:

csrwng commented:
@pweil yes it is possible to get an error in the output and in the progress writer we detect it as well... https://github.com/openshift/imagebuilder/blob/master/imageprogress/progress.go

btw, that's another example of something that parses the pull output.

Just a couple of comments on this pull...

1) why not use something like https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/credentialprovider/keyring.go to parse docker configs and narrow down which auths should be attempted.

2) (line 244) looks like after the first auth failure, you exit the function and never try other auths

pweil- commented:
> (line 244) looks like after the first auth failure, you exit the function and never try other auths

I totally missed that :1st_place_medal:

enoodle commented:
@pweil- I fixed your recent comments.
@csrwng about the authErr, I changed it to just `err` as authentication errors are no longer returned from there (This is why we are looking for them in the stream). About using the keyring, correct me I am wrong but thsi is for cases where the proccess lives longer. In our case "image-inspector" is called to pull one image and then goes away so it won't have a use for caching for the authentications.

csrwng commented:
@enoodle the keyring [will narrow down the list of auths for you to use based on your pull spec](https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/credentialprovider/keyring.go#L239); imho that seems better than just trying every auth entry from all your dockercfg's. You don't need to use that code exactly, but the approach seems better.

pweil- commented on pkg/inspector/image-inspector.go:
```diff
@@ -210,27 +226,30 @@ func (i *defaultImageInspector) pullImage(client *docker.Client) error {
+		go decodeDockerResponse(parsedErrors, reader)
+
+		if err := client.PullImage(imagePullOption, auth); err != nil {
+			return err
```
this return will exit the loop as @csrwng mentioned.  Other than that it looks like you've addressed all my asks.

Thanks!

enoodle replied:
This was authErr that I now changed to just `err` because authentication errors are only returned in the stream. Errors here should terminate the loop in my opinion because they are more general ones that will happen with any authentication.

enoodle commented:
@pweil- PTAL (hopefully last one? ) as I have dealt with the `client.PullImage` error and the "double" closing of the pipe by putting that call in a go routine and pushing that error to the errors channel.

enoodle commented:
about the keyring - it looks like what we should be doing, but I prefer adding it in a different PR to keep one subject here.

simon3z commented:
> latest impl looks really clean. LGTM I'm ok with a follow up issue for the keyring

@pweil- since I don't want to fix this more than once I asked @enoodle to look into the keyring now. If it's better and more reliable I'd prefer to go with that.
If not, then we'll go with this.

simon3z commented:
@enoodle it seems that this is way too large for the `2.1` release.

We'll commit the previous PR (the one @pweil- already acked) and we'll evaluate this for `2.2`.

enoodle commented:
@simon3z I separated to another PR.
